Fix part of #1633: Removed view dependency from ProfileEditViewModel

alokbharti opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
This PR fixes part of #1633 
Fixes part of #1870 

This PR contains following changes:
- ProfileEditViewModel had a reference to activity variable to set its title but now an existing livedata variable is used to expose the title value and then ProfileEditActivityPresenter observe that variable to set the title for ProfileEditActivity.
- Checked state of switch is now shifted to ProfileEditActivityPresenter
- Instead of using ViewModelProvider, now ProfileEditViewModel is directly injected in ProfileEditActivityPresenter
- Removed onRestoreInstanceState from ProfileEditActivity as viewmodel can survives configuration changes

Following tests are done to ensure that everything is working as it was before
- Created multiple users and verified that the profile name is visible as the title of the activity
- Verified the checked state of switch button in ProfileEditActivity
- Verified above after rotating screen multiple times
- Repeated above steps while Talkback is on
- Ran existing test file related to ProfileEditActivity

**Update**
- Fixed all the tests for ProfileEditActivity. (tested on pixel 2 emulator and Xioami Mi A3)

Roboelectric
![roboelectric](https://user-images.githubusercontent.com/26787263/97908707-0eb02e00-1d6d-11eb-8df4-e84b39bc6d8d.png)

Espresso
![espresso](https://user-images.githubusercontent.com/26787263/97908727-17086900-1d6d-11eb-9560-1e716ba43cb6.png)

## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

alokbharti committed:
Removed view dependency from ProfileEditViewModel

- Used MutableLiveData to store the value of ProfileEditActivity Title
  and then expose it using a livedata variable in profleEditViewModel
  class
- Observe the livedata variable to set ProfileEditActivity Title in
  ProfileEditActivityPresenter class

Arjupta commented:
@alokbharti you must ask for a review from any of the code maintainers else your PR will be noticed late. Like 
@rt4914 sir were viewing the original issue so you can ask him for review here.

alokbharti commented:
> @alokbharti you must ask for a review from any of the code maintainers else your PR will be noticed late. Like
> @rt4914 sir were viewing the original issue so you can ask him for review here.

@ARJUPTA I already know this but I was waiting for the checks to be finished.

alokbharti commented:
@rt4914 PTAL

anandwana001 commented on app/src/main/java/org/oppia/android/app/settings/profile/ProfileEditViewModel.kt:
```diff
@@ -27,6 +25,9 @@ class ProfileEditViewModel @Inject constructor(
 
   lateinit var profileName: String
 
+  private val _activityTitle = MutableLiveData<String>()
```
As per I understand from the issue and [this](https://github.com/oppia/oppia-android/blob/4c345db0f51c8a40870133113dd8ba7533a06d86/app/src/main/java/org/oppia/app/settings/profile/ProfileEditViewModel.kt) committed file, we have to work on `Switch`. 
@miaboloix Could you please correct me here?

rt4914 replied:
@anandwana001 you might need to approach @miaboloix directly w.r.t. this.

BenHenning replied:
@anandwana001 is correct, though there are actually two issues with this view model:
1. It relies on a view directly (the ``Switch``--view models should not depend on views)
2. It relies on Activity, but its lifecycle is managed by AndroidX's ViewModelProvider which means it will leak references to the previous activity during configuration changes

#1633 is meant to fix (1), but if possible we should also fix (2) while here. The way we've been fixing (2) everywhere else is just by injecting the view model directly into the presenter rather than using ViewModelProvider.

@alokbharti you'll need to figure out a way to make (1) work--we don't want to depend on ``Switch`` which may require introducing a callback interface or a live data from this view model.

alokbharti replied:
Hi @BenHenning , thanks for the suggestions. I'll fix those issues one by one for this viewmodel. For now, as suggested I'm doing the changes for the switch part.

BenHenning replied:
Thanks. While here, can you also do (2)? All that's needed is to switch the presenter from using ModelViewProvider to injecting the view model directly. There aren't straightforward tests that we can add for this, so we'll rely on existing tests to pass.

rt4914 commented:
Also, I have assigned this PR to @BenHenning  too as I am not completely sure about this solution yet.

BenHenning reviewed (changes requested):
Thanks @alokbharti! Left some thoughts.

alokbharti committed:
Resolved swith dependency from ProfileViewModel

- Added a livedata variable to expose the checked state of switch in
  ProfileViewModel through which ProfileEditActivityPresenter can
  observe and update the UI accordingly
- Removed switch variable initialization from viewmodel
- Added activity dependeny as it was before

alokbharti committed:
Added new lines as required

BenHenning commented on app/src/main/java/org/oppia/android/app/settings/profile/ProfileEditViewModel.kt:
```diff
@@ -23,7 +22,9 @@ class ProfileEditViewModel @Inject constructor(
-  private lateinit var switch: Switch
+
+  private val _isAllowedDownloadAccess = MutableLiveData<Boolean>()
+  val isAllowedDownloadAccess: LiveData<Boolean> = _isAllowedDownloadAccess
```
Public members of a class should have documentation. See https://developer.android.com/kotlin/style-guide#usage and https://developer.android.com/kotlin/style-guide#documentation.

BenHenning reviewed (approved):
Thanks @alokbharti! Left two comments. I noticed that your PR description is out-of-date--could you update that to reflect the changes in this PR?

Also to confirm, have you manually tested the app with these changes to make sure that the profile switch still works correctly?

Further, can you please also verify or add tests to make sure the switch functionality is working correctly after the changes from this PR?

BenHenning reviewed (changes requested):
Apologies, I meant to select 'request changes' in the last review.

alokbharti commented:
> Thanks @alokbharti! Left two comments. I noticed that your PR description is out-of-date--could you update that to reflect the changes in this PR?
> 
> Also to confirm, have you manually tested the app with these changes to make sure that the profile switch still works correctly?
> 
> Further, can you please also verify or add tests to make sure the switch functionality is working correctly after the changes from this PR?

Thanks @BenHenning , I've tested the app manually & using talkback and It's working correctly. I'll add the second part too in the next commit i.e. changes for activity title and will update the PR.

alokbharti committed:
Removed Switch & activity from ProfileViewModel

- Removed activity variable from ProfileViewModel, as we can simply use
  existing profile livedata variable to update the title of the
  ProfileEditActivity in the ProfileEditActivityPresenter
- Instead of using viewModelProvider in ProfileEditActivityPresenter to
  get viewmodel, now it is directly injected
- Since viewmodel can survive configuration changes, there is no need to
  create an extra variable to store profile name to use it in
  onRestoreInstanceState
- Added comments for public variable in ProfileEditViewModel

alokbharti committed:
Removed unused import

alokbharti committed:
Merge branch 'develop' into profileviewmodel-view-dependency

alokbharti commented:
Hi @BenHenning , I've updated my PR with the requested changes. PTAL

rt4914 commented:
> Hi @BenHenning , I've updated my PR with the requested changes. PTAL

@alokbharti Also, mention the reviewer in `Assignees` section too.

https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section

For now, I am doing that.

BenHenning commented on app/src/main/java/org/oppia/android/app/settings/profile/ProfileEditViewModel.kt:
```diff
@@ -14,18 +13,22 @@ import org.oppia.android.util.data.DataProviders.Companion.toLiveData
-  lateinit var profileName: String
+  private val _isAllowedDownloadAccess = MutableLiveData<Boolean>()
+
+  /**
+   * This variable is used to expose the switch state
+   * that determines whether the access to download is allowed or not
+   * to the presenter
+   **/
```
```suggestion
  /** Specifies whether download access has been enabled by the user. */
```

Couple things to note:
1. KDoc ends with ``*/`` not ``**/``
2. KDocs should end with punctuation
3. KDoc lines should be wrapped (not applicable in my suggestion since I simplified it)
4. Field documentation should describe _what_ the value represents, not who uses it and usually not how it's assigned (though in some cases this is important)

BenHenning reviewed (changes requested):
Thanks @alokbharti! Apologies for the late review.

I think the overall solution looks good. I did notice, however, that we don't actually have any tests for verifying the activity title in https://github.com/oppia/oppia-android/blob/develop/app/src/sharedTest/java/org/oppia/android/app/settings/profile/ProfileEditActivityTest.kt. Could you add one? I'm not yet sure why those tests are ``@Ignore``d (specifically why they aren't already passing), so if you run into problems with trying to get a new test added please let us know. I suggest just copying an existing test & adding the assertions for activity title to keep things simple.

alokbharti committed:
Removed @Ignore from activity title related testcases from ProfileEditActivityTest and modified description for the public variable in ProfielEditViewModel

alokbharti committed:
Merge branch 'develop' into profileviewmodel-view-dependency

alokbharti commented:
Hi @BenHenning , thanks for your review. I removed @Ignore annotation from activity title related tests but it seems that all those tests are passing in espresso but failing in Roboelectric. Can you help me here?

BenHenning reviewed (commented):
@alokbharti those tests were ignored because they are failing on Robolectric (see #973 for context). My comment earlier was not meant to suggest that you should remove the ``@Ignore``s (those tests are correctly ignored). What I meant earlier is I don't know specifically if any special work is needed to make ProfileEditActivity tests to work.

I suggest re-adding the ``@Ignore``s, and introducing a new test that is not ``@Ignore``d to test the activity title. That test should be possible in both Robolectric and Espresso.

Feel free to reach out on Gitter if you run into any issues.

anandwana001 reviewed (changes requested):
Adding a few points on doing testing and sending pr:

1. Check our wiki page [here](https://github.com/oppia/oppia-android/wiki/Oppia-Android-Testing#tips-to-run-test-cases-in-both-espresso-and-robolectric) for correctly running sharedTest in both espresso as well as robolectric.
2. Always add a screenshot showing all test cases pass/fail on both espresso and robolectric in PR description in order to help reviewer.
3. If there is any specific error, always add stack trace of it in code format.


For `ProfileEditActivityTest`, it will be simple to fix, you can check above mentioned wiki page and PR's of #973 

@rt4914 One concern I have here is, if you look at `layout-land/profile_edit_activity.xml`, this has `ScrollView` and it's portrait file not have any ScrollView. 
As per the tests, we have things like `onView(withId(R.id.profile_edit_allow_download_container)).perform(scrollTo())` which needs scrolling, but I didn't check running this espresso test on smaller devices where we truly need scrolling and will it crash or not. So, I think this might be one of the reason where robolectric fails. 

@alokbharti Could you check running espresso test on small devices and see if scrolling happens or not?

rt4914 commented:
Removing myself a reviewer as @BenHenning and @anandwana001 are already taking care of this PR.

anandwana001 commented:
> Adding a few points on doing testing and sending pr:
> 
> 1. Check our wiki page [here](https://github.com/oppia/oppia-android/wiki/Oppia-Android-Testing#tips-to-run-test-cases-in-both-espresso-and-robolectric) for correctly running sharedTest in both espresso as well as robolectric.
> 2. Always add a screenshot showing all test cases pass/fail on both espresso and robolectric in PR description in order to help reviewer.
> 3. If there is any specific error, always add stack trace of it in code format.
> 
> For `ProfileEditActivityTest`, it will be simple to fix, you can check above mentioned wiki page and PR's of #973
> 
> @rt4914 One concern I have here is, if you look at `layout-land/profile_edit_activity.xml`, this has `ScrollView` and it's portrait file not have any ScrollView.
> As per the tests, we have things like `onView(withId(R.id.profile_edit_allow_download_container)).perform(scrollTo())` which needs scrolling, but I didn't check running this espresso test on smaller devices where we truly need scrolling and will it crash or not. So, I think this might be one of the reason where robolectric fails.
> 
> @alokbharti Could you check running espresso test on small devices and see if scrolling happens or not?

@alokbharti Any update on this?

alokbharti commented:
> @alokbharti Any update on this?

Hi @anandwana001 I'm working on fixing tests that are failing on roboelectric #1870 . I'm getting some error related to the tests that are using `scrollTo()` method. Will commit my changes soon.

anandwana001 commented:
> > @alokbharti Any update on this?
> 
> Hi @anandwana001 I'm working on fixing tests that are failing on roboelectric #1870 . I'm getting some error related to the tests that are using `scrollTo()` method. Will commit my changes soon.

ok, Thanks for the update. Also, please update your PR with the lastest develop.

alokbharti committed:
Resolved tests that are failing in roboelectric except that uses scrollTo() method

alokbharti committed:
Merge branch 'develop' into profileviewmodel-view-dependency

alokbharti committed:
FirebaseApp initializtion was giving an IO exception and since it is not used in activity, removed it from the test

alokbharti committed:
Removed unused imports

alokbharti commented:
Hi @BenHenning @anandwana001 , I've updated my PR. PTAL.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/settings/profile/ProfileEditActivityTest.kt:
```diff
@@ -294,7 +299,10 @@ class ProfileEditActivityTest {
     ).use {
       onView(withId(R.id.profile_delete_button)).perform(click())
-      onView(withText(R.string.profile_edit_delete_dialog_positive)).perform(click())
+      onView(withText(context.getString(R.string.profile_edit_delete_dialog_positive)))
```
Why isn't ``withText(stringResId)`` sufficient here & below?

alokbharti replied:
Yes, `withText(stringResId)` is sufficient here & below. Actually, I followed some other PRs and made this change. Will update soon. Thanks! : )

rt4914 replied:
Resolving this on behalf of @BenHenning

BenHenning reviewed (commented):
Thanks @alokbharti for fixing the tests in ProfileEditActivityTest. That actually wasn't needed per my earlier comment, but if they're passing now that's really appreciated. Can you please update the PR description to mention that this PR is also partially fixing #973?

Beyond that, please look again at https://github.com/oppia/oppia-android/pull/1925#pullrequestreview-507899385. I'm requesting that you add new tests, and unless I'm mistaken I don't think any new tests were added yet for ProfileEditActivityTest. Please let me know if that comment isn't clear and I will try to provide more detail.

alokbharti commented:
> Thanks @alokbharti for fixing the tests in ProfileEditActivityTest. That actually wasn't needed per my earlier comment, but if they're passing now that's really appreciated. Can you please update the PR description to mention that this PR is also partially fixing #973?
> 
> Beyond that, please look again at [#1925 (review)](https://github.com/oppia/oppia-android/pull/1925#pullrequestreview-507899385). I'm requesting that you add new tests, and unless I'm mistaken I don't think any new tests were added yet for ProfileEditActivityTest. Please let me know if that comment isn't clear and I will try to provide more detail.

Thanks @BenHenning for your review. Just have a doubt. Since tests related to the activity title are already there in ProfileEditActivityTest class and now they are passing on both (espresso and roboelectric), should I add more tests for the same? Also, there is a separate issue (#1870) created for fixing roboelectric tests for this class, so I'm updating the description to mention that too.

anandwana001 commented:
@alokbharti This PR is fixing only a part of #1870 and that too partially as I can see there are some other issues which is left to resolve under #1870

alokbharti commented:
> @alokbharti This PR is fixing only a part of #1870 and that too partially as I can see there are some other issues which is left to resolve under #1870

@anandwana001 Yes, you're right. I'm trying to fix all of the tests but got stuck on the tests that uses `scrollTo()` method.
This is the error I'm getting:

````
androidx.test.espresso.PerformException: Error performing 'scroll to' on view 'with id: org.oppia.android:id/profile_reset_button'.

	at androidx.test.espresso.PerformException$Builder.build(PerformException.java:86)
	at androidx.test.espresso.base.DefaultFailureHandler.getUserFriendlyError(DefaultFailureHandler.java:87)
	at androidx.test.espresso.base.DefaultFailureHandler.handle(DefaultFailureHandler.java:59)
	at androidx.test.espresso.ViewInteraction.waitForAndHandleInteractionResults(ViewInteraction.java:322)
	at androidx.test.espresso.ViewInteraction.desugaredPerform(ViewInteraction.java:178)
	at androidx.test.espresso.ViewInteraction.perform(ViewInteraction.java:119)
	at org.oppia.android.app.settings.profile.ProfileEditActivityTest.testProfileEditActivity_configurationChange_startActivityWithUserProfile_clickResetPin_checkOpensProfileResetPinActivity(ProfileEditActivityTest.kt:242)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.robolectric.RobolectricTestRunner$HelperTestRunner$1.evaluate(RobolectricTestRunner.java:546)
	at org.robolectric.internal.SandboxTestRunner$2.lambda$evaluate$0(SandboxTestRunner.java:252)
	at org.robolectric.internal.bytecode.Sandbox.lambda$runOnMainThread$0(Sandbox.java:89)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.RuntimeException: Action will not be performed because the target view does not match one or more of the following constraints:
(view has effective visibility=VISIBLE and is descendant of a: (is assignable from class: class android.widget.ScrollView or is assignable from class: class android.widget.HorizontalScrollView or is assignable from class: class android.widget.ListView))
Target view: "AppCompatButton{id=2131296792, res-name=profile_reset_button, visibility=VISIBLE, width=960, height=144, has-focus=false, has-focusable=true, has-window-focus=true, is-clickable=true, is-enabled=true, is-focused=false, is-focusable=true, is-layout-requested=false, is-selected=false, layout-params=android.widget.LinearLayout$LayoutParams@3da69362, tag=null, root-is-layout-requested=false, has-input-connection=false, x=0.0, y=150.0, text=Reset PIN, input-type=0, ime-target=false, has-links=false}"
	at androidx.test.espresso.ViewInteraction.doPerform(ViewInteraction.java:252)
	at androidx.test.espresso.ViewInteraction.access$100(ViewInteraction.java:65)
	at androidx.test.espresso.ViewInteraction$1.call(ViewInteraction.java:158)
	at androidx.test.espresso.ViewInteraction$1.call(ViewInteraction.java:155)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at android.os.Handler.$$robo$$android_os_Handler$handleCallback(Handler.java:873)
	at android.os.Handler.handleCallback(Handler.java)
	at android.os.Handler.$$robo$$android_os_Handler$dispatchMessage(Handler.java:99)
	at android.os.Handler.dispatchMessage(Handler.java)
	at org.robolectric.shadows.ShadowPausedLooper$IdlingRunnable.run(ShadowPausedLooper.java:308)
	at org.robolectric.shadows.ShadowPausedLooper.executeOnLooper(ShadowPausedLooper.java:273)
	at org.robolectric.shadows.ShadowPausedLooper.idle(ShadowPausedLooper.java:85)
	at org.robolectric.android.internal.LocalControlledLooper.drainMainThreadUntilIdle(LocalControlledLooper.java:15)
	at androidx.test.espresso.ViewInteraction.waitForAndHandleInteractionResults(ViewInteraction.java:318)
	... 20 more
```

anandwana001 commented:
@alokbharti any update on this?

alokbharti commented:
> @alokbharti any update on this?

Hi @anandwana001 , need help to solve the issue that I mentioned [above](https://github.com/oppia/oppia-android/pull/1925#issuecomment-715064412)

anandwana001 commented:
@alokbharti Let's connect over gitter or plan a meet for further discussion. 
All we need to discuss is about the delay in robolectric test cases using TestCoroutineDispatchers.

alokbharti committed:
added scrollview in port layout of profile_edit_activity to resolve roboelectric tests

alokbharti committed:
Removed unused imports

alokbharti committed:
Merge branch 'develop' into profileviewmodel-view-dependency

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/settings/profile/ProfileEditActivityTest.kt:
```diff
@@ -237,14 +235,14 @@ class ProfileEditActivityTest {
       onView(isRoot()).perform(orientationLandscape())
+      testCoroutineDispatchers.runCurrent()
       onView(withId(R.id.profile_reset_button)).perform(scrollTo()).perform(click())
+      testCoroutineDispatchers.runCurrent()
```
I don't think we need a delay here.

alokbharti replied:
DONE

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/settings/profile/ProfileEditActivityTest.kt:
```diff
@@ -237,14 +235,14 @@ class ProfileEditActivityTest {
       )
     ).use {
       onView(isRoot()).perform(orientationLandscape())
+      testCoroutineDispatchers.runCurrent()
```
I don't this test case needs delay!!

alokbharti replied:
DONE

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/settings/profile/ProfileEditActivityTest.kt:
```diff
@@ -263,18 +262,19 @@ class ProfileEditActivityTest {
         1
       )
     ).use {
+      testCoroutineDispatchers.runCurrent()
```
Same here, no need for delay in this test case.

alokbharti replied:
DONE

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/settings/profile/ProfileEditActivityTest.kt:
```diff
@@ -294,31 +292,35 @@ class ProfileEditActivityTest {
         1
       )
     ).use {
+      testCoroutineDispatchers.runCurrent()
```
this delay is not required

alokbharti replied:
DONE

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/settings/profile/ProfileEditActivityTest.kt:
```diff
@@ -294,31 +292,35 @@ class ProfileEditActivityTest {
       onView(isRoot()).perform(orientationLandscape())
       onView(withId(R.id.profile_delete_button)).perform(scrollTo()).perform(click())
-      onView(withText(R.string.profile_edit_delete_dialog_positive)).perform(click())
+      testCoroutineDispatchers.runCurrent()
```
this delay is not required

alokbharti replied:
DONE

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/settings/profile/ProfileEditActivityTest.kt:
```diff
@@ -90,16 +90,20 @@ class ProfileEditActivityTest {
     setUpTestApplicationComponent()
+    testCoroutineDispatchers.registerIdlingResource()
     profileTestHelper.initializeProfiles()
-    FirebaseApp.initializeApp(context)
```
@Sarthak2601 Could you provide some context on this, as could we remove these lines or not? 
Do we require to init firebase app?

alokbharti replied:
It was giving IOException while running tests on roboelectric and besides I don't see any need to init firebase here as we aren't using it anywhere for ProfileEditActivity or it's test class

Sarthak2601 replied:
I'm not sure if the tests would run without it as it was put there to ensure that Firebase related dependencies(available via DI) are initialised. However, if the tests are running fine, there's no need to initialise it here.

alokbharti replied:
Tests are running fine without initializing firebase

anandwana001 commented on app/src/main/java/org/oppia/android/app/settings/profile/ProfileEditActivityPresenter.kt:
```diff
@@ -21,12 +19,10 @@ import javax.inject.Inject
-  private val editViewModel: ProfileEditViewModel by lazy {
-    getProfileEditViewModel()
-  }
+  @Inject
```
Leave an empty line above

alokbharti replied:
DONE

anandwana001 commented on app/src/main/java/org/oppia/android/app/settings/profile/ProfileEditActivityPresenter.kt:
```diff
@@ -21,12 +19,10 @@ import javax.inject.Inject
-    getProfileEditViewModel()
-  }
+  @Inject
+  lateinit var editViewModel: ProfileEditViewModel
```
Rename to `profileEditViewModel`

alokbharti replied:
DONE

anandwana001 commented on app/src/main/java/org/oppia/android/app/settings/profile/ProfileEditActivityPresenter.kt:
```diff
@@ -38,9 +34,9 @@ class ProfileEditActivityPresenter @Inject constructor(
     editViewModel.setProfileId(
-      profileId,
-      activity.findViewById<Switch>(R.id.profile_edit_allow_download_switch)
+      profileId
```
`editViewModel.setProfileId(profileId)` it should be in single line

alokbharti replied:
DONE

anandwana001 commented on app/src/main/java/org/oppia/android/app/settings/profile/ProfileEditActivityPresenter.kt:
```diff
@@ -64,6 +60,20 @@ class ProfileEditActivityPresenter @Inject constructor(
       showDeletionDialog(profileId)
     }
 
+    editViewModel.profile.observe(
```
This looks more clean!! What's you though on this? Same for the below one. 
```kotlin
editViewModel.profile.observe(
      activity, Observer {
        activity.title = it.name
      }
    )
```

alokbharti replied:
Lint test fails here due to this.

anandwana001 commented on app/src/main/res/layout/profile_edit_activity.xml:
```diff
@@ -12,228 +12,235 @@
   </data>
 
-  <androidx.constraintlayout.widget.ConstraintLayout
+  <ScrollView
```
Do the `Reformat Code`. Check our wiki guide here - https://github.com/oppia/oppia-android/wiki/Ktlint-Guide#macros
We will be updating the wiki specifying for XML code too., as of now it says for kotlin only, but do the same for XML.

alokbharti replied:
DONE

anandwana001 reviewed (changes requested):
Thanks @alokbharti 
Check my comments below.

rt4914 commented:
@alokbharti Please apply changes suggested by @anandwana001

alokbharti committed:
Added feedback changes

- Removed unneccessary use of testCoroutineDispatchers.runCurrent()
- Renamed editViewModel to profileEditViewModel
- Few Indentation

alokbharti committed:
modified indentation

alokbharti commented:
@anandwana001 @rt4914 , did the changes. PTAL.

rt4914 commented on app/src/main/java/org/oppia/android/app/settings/profile/ProfileEditViewModel.kt:
```diff
@@ -14,18 +13,18 @@ import org.oppia.android.util.data.DataProviders.Companion.toLiveData
-  private lateinit var switch: Switch
 
-  lateinit var profileName: String
+  private val _isAllowedDownloadAccess = MutableLiveData<Boolean>()
```
This seems like a convention that we haven't been following in the code.
1. Has any reviewer suggested this change? If so, please mention the comment link.
2. If no one has suggested any comment related to this, I suggest you renaming to something else like
`isAllowedDownloadAccessLiveData`.

alokbharti replied:
No one has suggested this convention. Renamed to `isAllowedDownloadAccessMutableLiveData` as there is already an another livedata variable with same name.

rt4914 commented on app/src/sharedTest/java/org/oppia/android/app/settings/profile/ProfileEditActivityTest.kt:
```diff
@@ -334,13 +331,12 @@ class ProfileEditActivityTest {
   @Test
-  // TODO(#973): Fix ProfileEditActivityTest
-  @Ignore
   fun testProfileEditActivity_configurationChange_startActivityWithUserHasDownloadAccess_checkSwitchIsChecked() { // ktlint-disable max-line-length
```
We should remove `ktlint` comment by making the naming within the character limit.

Suggestion:
`fun testProfileEdit_configurationChange_startWithUserHasDownloadAccess_checkSwitchIsChecked() {`

alokbharti replied:
DONE

rt4914 commented on app/src/sharedTest/java/org/oppia/android/app/settings/profile/ProfileEditActivityTest.kt:
```diff
@@ -294,14 +289,15 @@ class ProfileEditActivityTest {
   @Test
-  // TODO(#973): Fix ProfileEditActivityTest
-  @Ignore
   fun testProfileEditActivity_configurationChange_startActivityWithUserProfile_clickProfileDeletionButton_clickDelete_checkReturnsToProfileListActivity() { // ktlint-disable max-line-length
```
Suggestion 
`fun testProfileEdit_configChange_startAnyProfile_deleteTheProfile_checkReturnsToProfileList() {`

alokbharti replied:
DONE

rt4914 commented on app/src/sharedTest/java/org/oppia/android/app/settings/profile/ProfileEditActivityTest.kt:
```diff
@@ -284,8 +281,6 @@ class ProfileEditActivityTest {
   @Test
-  // TODO(#973): Fix ProfileEditActivityTest
-  @Ignore
   fun testProfileEditActivity_startActivityWithUserProfile_clickProfileDeletionButton_clickDelete_checkReturnsToProfileListActivity() { // ktlint-disable max-line-length
```
Suggestion:
`fun testProfileEdit_startWithUserProfile_deleteProfile_checkReturnsToProfileList() {`

alokbharti replied:
DONE

rt4914 commented on app/src/sharedTest/java/org/oppia/android/app/settings/profile/ProfileEditActivityTest.kt:
```diff
@@ -263,8 +260,6 @@ class ProfileEditActivityTest {
   @Test
-  // TODO(#973): Fix ProfileEditActivityTest
-  @Ignore
   fun testProfileEditActivity_configurationChange_startActivityWithUserProfile_clickProfileDeletionButton_checkOpensDeletionDialog() { // ktlint-disable max-line-length
```
Suggestion
`fun testProfileEdit_configChange_startWithUserProfile_clickDelete_checkOpensDeletionDialog() {`

alokbharti replied:
DONE

rt4914 commented on app/src/sharedTest/java/org/oppia/android/app/settings/profile/ProfileEditActivityTest.kt:
```diff
@@ -243,8 +241,6 @@ class ProfileEditActivityTest {
   @Test
-  // TODO(#973): Fix ProfileEditActivityTest
-  @Ignore
   fun testProfileEditActivity_startActivityWithUserProfile_clickProfileDeletionButton_checkOpensDeletionDialog() { // ktlint-disable max-line-length
```
Suggestion
`fun testProfileEdit_startWithUserProfile_clickProfileDeletionButton_checkOpensDeletionDialog() { `

alokbharti replied:
DONE

rt4914 commented on app/src/sharedTest/java/org/oppia/android/app/settings/profile/ProfileEditActivityTest.kt:
```diff
@@ -227,8 +227,6 @@ class ProfileEditActivityTest {
   @Test
-  // TODO(#973): Fix ProfileEditActivityTest
-  @Ignore
   fun testProfileEditActivity_configurationChange_startActivityWithUserProfile_clickResetPin_checkOpensProfileResetPinActivity() { // ktlint-disable max-line-length
```
Suggestion
`fun testProfileEdit_configurationChange_startAnyProfile_clickResetPin_checkOpensResetPin() {`

alokbharti replied:
DONE

rt4914 commented on app/src/sharedTest/java/org/oppia/android/app/settings/profile/ProfileEditActivityTest.kt:
```diff
@@ -227,8 +227,6 @@ class ProfileEditActivityTest {
   }
 
```
Suggestion at line `217`
`fun testProfileEdit_startActivityWithUserProfile_clickResetPin_checkOpensProfileResetPin() {`

alokbharti replied:
DONE

rt4914 reviewed (changes requested):
I have suggestion changes.

There are some changes related to test names. The main concept is to change the name such that we can remove `// ktlint-disable max-line-length`.
You can do this by following changes.
Remove `Activity` word from the test-name.
Change `configurationChange` to `configChange`.

alokbharti committed:
Added PR requested changes

- Removed '// ktlint-disable max-line-length' by following suggested
  convention
- Renamed _isAllowedDownloadAccess to
  isAllowedDownloadAccessMutableLiveData

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/settings/profile/ProfileEditActivityTest.kt:
```diff
@@ -90,16 +90,20 @@ class ProfileEditActivityTest {
   @Inject
   lateinit var profileManagementController: ProfileManagementController
 
+  @Inject
```
add a top-level comment to class. 
`/** Tests for [StateFragment]. */`
Check `StateFragmentTest` for reference.

alokbharti replied:
DONE

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/settings/profile/ProfileEditActivityTest.kt:
```diff
@@ -143,15 +145,14 @@ class ProfileEditActivityTest {
     ActivityScenario.launch<ProfileEditActivity>(
       ProfileEditActivity.createProfileEditActivity(
         context,
         1
```
let's use named arguments here and all places.

alokbharti replied:
DONE

anandwana001 replied:
It doesn't seem to be done!! Could you please check once?

alokbharti replied:
I checked, it is done in my last commit.

anandwana001 replied:
<img width="613" alt="Screenshot 2020-11-05 at 08 54 26" src="https://user-images.githubusercontent.com/25057618/98194055-87b5ae00-1f44-11eb-9979-c61ac7da9e2a.png">

I didn't find the changes. Checked the latest commit.

anandwana001 replied:
named arguments -> 
`profileId = 1`

alokbharti replied:
Oh sorry, I misunderstood. Will do shortly

alokbharti replied:
DONE

anandwana001 reviewed (changes requested):
nit changes @alokbharti

alokbharti committed:
Added feedback changes, renaming tests

alokbharti committed:
merged develop and resolved conflict

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/settings/profile/ProfileEditActivityTest.kt:
```diff
@@ -134,15 +132,14 @@ class ProfileEditActivityTest {
-  @Ignore
-  fun testProfileEditActivity_startActivityWithAdminProfile_checkAdminInfoIsDisplayed() {
+  fun testProfileEdit_startWithAdminProfile_checkAdminInfoIsDisplayed() {
     ActivityScenario.launch<ProfileEditActivity>(
```
One last nit change, else it's good to go. 
Rather than import `ActivityScenario` as a whole, we only need `launch`.

```diff
- import androidx.test.core.app.ActivityScenario
+ import androidx.test.core.app.ActivityScenario.launch
```

alokbharti replied:
DONE

anandwana001 reviewed (changes requested):
nit change.

alokbharti committed:
Used launch instead of ActivityScenario.launch

alokbharti committed:
Added named argument

anandwana001 reviewed (approved):
LGTM, Thanks @alokbharti

rt4914 reviewed (approved):
LGTM, thanks.

rt4914 commented:
@anandwana001 Please resolve your comment threads. Also if you find any comment which I have not resolved feel free to add me again.

@alokbharti We will have to wait for @BenHenning review on this as the merging is blocked. Mostly by End of this month.

BenHenning commented on app/src/main/res/layout/profile_edit_activity.xml:
```diff
@@ -12,228 +12,236 @@
     android:layout_height="match_parent"
-    android:background="@color/profileEditBackground"
-    android:clipToPadding="false">
+    android:layout_margin="0dp"
```
Isn't margin 0dp by default? Why is this needed?

alokbharti replied:
Removed.

BenHenning commented on app/src/main/res/layout/profile_edit_activity.xml:
```diff
@@ -12,228 +12,236 @@
-    android:clipToPadding="false">
+    android:layout_margin="0dp"
+    android:fillViewport="true"
+    android:padding="0dp">
```
Ditto for padding.

alokbharti replied:
Removed

BenHenning commented on app/src/main/res/layout/profile_edit_activity.xml:
```diff
@@ -12,228 +12,236 @@
+      android:clipToPadding="false">
 
       <LinearLayout
+        android:id="@+id/linearLayout"
```
While you're here, can you fix this ID to either be more descriptive about what it contains (see other layout IDs for references), or remove it if nothing's referencing the ID?

alokbharti replied:
changed to profile_edit_actions_container

BenHenning reviewed (approved):
Thanks @alokbharti! Really nice improvements to the tests.

Left a couple of comments, but the PR overall LGTM.

alokbharti committed:
Removed 0dp margin and padding from scrollview and renamed id for linearlayout

rt4914 commented:
@alokbharti If you are looking for another approval from Ben. Please assign him this PR. Thanks.

BenHenning reviewed (approved):
Thanks @alokbharti, LGTM!

BenHenning commented:
@Sarthak2601 I'm not seeing your comments for some reason, so going ahead and merging this. Please follow-up with an issue if you have any additional requests for this implementation.
